rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,205 @@
+# Composable Pipelines
+
+- Authors:
+  - [Matthias Vallentin](https://github.com/mavam)
+
+## Overview
+
+This proposal ideates a new UX around VAST's pipelines. The fundamental idea is
+that *everything is a pipeline* and build a composable UX around this concept.
+
+## Problem Statement
+
+Pipelines currently are only deployable at [predefined
+points](https://vast.io/docs/use-vast/transform). For example, they can be
+triggered for specific events on import/export, or invoked by compaction.
+
+As we are bringing pipelines to the query language, we need to solve a UX
+challenge: how do we make the data flow explicit and provide a unified
+experience, especially in the presence of custom plugins?
+
+## Solution Proposal
+
+Assume for a moment everything is a pipeline operator. Then we may write:
+
+ ```bash
+vast 'from s3://aws |
+      read json |
+      group-by src |
+      summarize count(dst) |
+      write feather |
+      to /path/to/file.feather'
+```
+
+Just to illustrate the composability, we could take this to an extreme when
+performance doesn't matter:
+
+```bash
+vast from s3://aws |
+  vast read json |
+  vast group-by |
+  vast summarize |
+  vast write feather |
+  vast to /path/to/file.feather
+```
+
+But let's take a step back and just assume that we add a new `exec` command that
+just executes a pipeline.
+
+Then we can model the ingestion as follows: load data via stdin, push into a
+parser, then send off to a remote VAST node. For example, we would rewrite
+`import zeek` as:
+
+```bash
+vast exec 'from - | parse zeek | to vast://1.2.3.4'
+```
+
+Likewise, `export json EXPRESSION` would become:
+
+```bash
+vast exec 'from vast://1.2.3.4 |
+           where EXPRESSION |
+           write json |
```

Comment:
The only constraint would be that input and output must be compatible types.

Let `Op1<In1, Out1>` and `Op2<In2, Out2>` be two operators. They can be chained to `Op1 | Op2` iff `Out1 == In2`.

In theory, there could be a weaker constraint, i.e., `Out1` must be only *convertible to* `In2`. But we'd probably start with equality for KISS reasons.

This does not apply to the ends of the pipeline, the sources and sinks, because they have a side effect on one side.

Reply:
(Tangent, happy to discuss elsewhere and post a summary here)

So, about the sources, just had a use-case in the last few days. How would could I "join" Suricata alerts and Zeek data in VAST? Assuming I have the community IDs in both sources, I want to output sets of (Suricata alert, Zeek logs). Currently I would looking at the Suricata alerts with one command, picking an interesting one and issuing another command with the community ID. That's maybe more command than I want to use, especially if in the end the Zeek logs reveal that the alerts wasn't interesting and there a many more Suricata alerts to investigate. Such capability could help spotting noisy Suricata alerts false positives in this example.

I thought that maybe VAST pipelines could provide (parts of) a solution. :-)

Reply:
It's a great use case. We are working the exact same thing again. :-)

Because a join is a hard operation to implement, we recast this problem: *"given a Suricata alerts, find me all related Zeek logs"*.

It turns out we already have an operator for this. Unfortunately it never made it beyond a special case for Suricata and PCAP. It's called `pivot`.

Today, you can write this:

```bash
vast pivot pcap.packet '#type == "suricata.alert" && 10.0.0.0/8'
```

This creates a two-step:
1. Produce all the records of the expression.
2. Look in every record for a field `community_id`, extract the value, and run a new query `#type == "pcap.packet" && community_id == X`

The end result is that you get a PCAP trace for a subset of your alerts. You can imagine how that translates to Zeek logs.

In pipeline land, `pivot` would be an operator with a side effect that pulls in more data through another query/pipeline. Precisely like a join. Here's how it would look like:

```bash
vast pull 'where "#type == "suricata.alert" |
           pivot community_id'
```

If we had the ability to extend the linear pipeline to a DAG with multiple edges, we would need more powerful constructs. Here's a `join` strawman:

```bash
vast pull 'fuse (
             => where #type == "suricata.alert"
             => where #field == "community_id" && #type != "suricata.alert"
           ) | right join on community_id
```

It would be joining Suricata alerts with all other types that have a field `community_id`.

This is a bit much to chew off for us right now, though. A `join` operator will probably be its own proposal.

Reply:
:100: Nice! Looks forward to that future proposal as well ;-) It's nice to see everything coming together and in relatively easy to use way as a user. Feel free to close the thread.